RiotSwiftUI/Modules/Authentication/Common/Service/MatrixSDK/RegistrationWizard.swift

```diff
@@ -0,0 +1,245 @@
+        }
+        
+        let parameters = RegistrationParameters(auth: AuthenticationParameters(type: kMXLoginFlowTypeDummy, session: session))
+        return try await performRegistrationRequest(parameters: parameters)
+    }
+
+    /// Perform the "m.login.email.identity" or "m.login.msisdn" stage.
+    ///
+    /// - Parameter threePID the threePID to add to the account. If this is an email, the homeserver will send an email
+    /// to validate it. For a msisdn a SMS will be sent.
+    func addThreePID(threePID: RegisterThreePID) async throws -> RegistrationResult {
+        pendingData.currentThreePIDData = nil
+        return try await sendThreePID(threePID: threePID)
+    }
+
+    /// Ask the homeserver to send again the current threePID (email or msisdn).
+    func sendAgainThreePID() async throws -> RegistrationResult {
+        guard let threePID = pendingData.currentThreePIDData?.threePID else {
+            throw AuthenticationError.createAccountNotCalled
+        }
+        return try await sendThreePID(threePID: threePID)
+    }
+
+    /// Send the code received by SMS to validate a msisdn.
+    /// If the code is correct, the registration request will be executed to validate the msisdn.
+    func handleValidateThreePID(code: String) async throws -> RegistrationResult {
+        return try await validateThreePid(code: code)
+    }
+
+    /// Useful to poll the homeserver when waiting for the email to be validated by the user.
+    /// Once the email is validated, this method will return successfully.
+    /// - Parameter delay How long to wait before sending the request.
+    func checkIfEmailHasBeenValidated(delay: TimeInterval) async throws -> RegistrationResult {
+        guard let parameters = pendingData.currentThreePIDData?.registrationParameters else {
+            throw AuthenticationError.noPendingThreePID
+        }
+
+        return try await performRegistrationRequest(parameters: parameters, delay: delay)
+    }
+    
+    // MARK: - Private
+    
+    private func validateThreePid(code: String) async throws -> RegistrationResult {
+        guard let threePIDData = pendingData.currentThreePIDData else {
+            throw AuthenticationError.noPendingThreePID
+        }
+        
+        guard let url = threePIDData.registrationResponse.submitURL else {
+            throw AuthenticationError.missingThreePIDURL
+        }
+        
+        
+        let validationBody = ThreePIDValidationCodeBody(clientSecret: pendingData.clientSecret,
+                                                        sessionID: threePIDData.registrationResponse.sessionID,
+                                                        code: code)
+        let validationDictionary = try validationBody.dictionary()
+        
+        #warning("Seems odd to pass a nil baseURL and then the url as the path, yet this is how MXK3PID works")
+        guard let httpClient = MXHTTPClient(baseURL: nil, andOnUnrecognizedCertificateBlock: nil) else {
```

Comment:
Could the code below be moved to the `RestClient`?

Reply:
Thanks, this also lead me to discover that `@Body params: validationDictionary` in Kotlin actually meant it goes in the body and not in the parameters. Would've found it out soon enough I guess, but its saved me from debugging it!